Add electric kiwi integration

mikey0000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
Addition of the Electric Kiwi integration, adding first (since its my first integration and time contributing here) a basic sensor that provides account balance, once this is merged I'll PR the additional sensors I've written. I've also written the associated dependency electrickiwi-api


## Type of change
Integration - Electric Kiwi, a utility (power) provider in New Zealand

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

-  This PR fixes or closes issue: fixes #
-  This PR is related to issue:
-  Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25332

Keys added by NabuCasa for production and staging from Electric Kiwi, so production ready.

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mikey0000 committed:
first start at porting existing utility-like plugin to work with electric kiwi

basic balance sensor for Electric Kiwi

add in logger and a few other possibly required imports

pinning dependencies and adding integration type

Fix copy/paste documentation errors

Add in auth2 flow and make updates to various flows

fix a few PR issues raised with duplicate consts and an import and an issue with basic auth

Add unique id and fix issue with initial call for connection id and customer number

Use NAME const and remove commented out code and consts

Move scope values to const

Tests pass for auth flow

Sort out tests, owners, coverage and rebasing

change to production url

add in missing reauth strings

if somehow the token is valid but the request fails raise catch the error and re-auth on set active session

move to async forward entry setups as per warning

use correct account balance

make the entities easier to extend and add one more

Add a few more entities and correctly classify them into their classes

no need for private async update now

24 hours is a bit too long, every 6 hours seems more reasonable for keeping it up to date.

move to using data coordinator pattern

move the coordinator to its own file

add select and seperate sensor entity for the start time of HOP

update manifest with latest version of the API

add electric kiwi to the strict and mypy configs

add in end time and tidy up what unique id returns and bump version of api for model fixes

address the checklist and remove account balance over hop as hop is more useful

address mypy issues

address comments and a few more tidy ups, accidentally removed async entry so restored that too

remove homekit item from manifest

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

remove unneeded items from manifest

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

move to better pattern for updating coordinator data manually

fix trailing comma

fix model duplicate names and tidy up sensor function, no need for ignores

matthuisman commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,75 @@
+"""Support for Flick Electric Pricing data."""
```
copy/paste error :) EK != Flick

matthuisman commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,75 @@
+async def async_setup_entry(
+        hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Flick Sensor Setup."""
```
Electric Kiwi Sensor Setup

mikey0000 commented:
Nice catch I'll sort that out.

2/11/2022 10:41:32 am Matt Huisman ***@***.***>:

> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In homeassistant/components/electric_kiwi/sensor.py[https://github.com/home-assistant/core/pull/81149#discussion_r1010925178]:
> 
>> +
> +from .const import DOMAIN
> +from ...util import Throttle
> +
> +_LOGGER = logging.getLogger(__name__)
> +
> +SCAN_INTERVAL = timedelta(hours=24)
> +
> +ATTRIBUTION = "Data provided by Electric Kiwi Juice Hacker API"
> +FRIENDLY_NAME = "Electric Kiwi Account Balance"
> +
> +
> +async def async_setup_entry(
> +        hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
> +) -> None:
> +    """Flick Sensor Setup."""
> 
> Electric Kiwi Sensor Setup
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/home-assistant/core/pull/81149#pullrequestreview-1164083316], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AALDOPNCXELOODDWJD2JJILWGGEPXANCNFSM6AAAAAARRIYCNE].
> You are receiving this because you authored the thread.[Tracking image][https://github.com/notifications/beacon/AALDOPMAUNFSHAUOYZ2QY3DWGGEPXA5CNFSM6AAAAAARRIYCNGWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSFMKAHI.gif]
>

ZephireNZ commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,75 @@
+
+SCAN_INTERVAL = timedelta(hours=24)
+
+ATTRIBUTION = "Data provided by Electric Kiwi Juice Hacker API"
```
This is defined twice, here and in `const`

ZephireNZ commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,75 @@
+    async_add_entities([ElectricKiwiBalanceSensor(api)], True)
+
+
+class ElectricKiwiBalanceSensor(SensorEntity):
```
I would recommend adding a `unique_id` to this sensor, so it can be renamed etc in the UI.

Could probably use account number for that.

mikey0000 replied:
Can you point me to the documentation for this? Yes account number would be good but I can only retrieve that after setup, so not sure how easy that is.

mikey0000 replied:
nevermind sorted it out

ZephireNZ commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,75 @@
+from electrickiwi_api.model import AccountBalance
+
+from .const import DOMAIN
+from ...util import Throttle
```
This import looks a bit odd, can probably just be `homeassistant.util`

ZephireNZ commented on homeassistant/components/electric_kiwi/const.py:
```diff
@@ -0,0 +1,11 @@
+"""Constants for the Electric Kiwi integration."""
+
+NAME = "Electric Kiwi"
+DOMAIN = "electric_kiwi"
+DOMAIN_DATA = f"{DOMAIN}_data"
+VERSION = "0.0.1"
+ATTRIBUTION = "Data provided by the Juice Hacker API"
+ISSUE_URL = "https://github.com/mikey0000/core/issues"
```
`NAME`, `DOMAIN_DATA`, `VERSION`, and `ISSUE_URL` are being defined but not actually used anywhere.

mikey0000 replied:
Thanks for the constructive feedback, yeah not merge ready yet and appreciate the eyes over it anyway. Was hoping to get some feedback on the Oauth process mostly then tidy up. But after finding their documentation and actual function for refreshing is different I need to go off and create my own refresh process that matches theirs.

ZephireNZ reviewed (approved):
LGTM 👍

mikey0000 commented:
Thank you!

2/01/2023 3:41:40 pm Brynley McDonald ***@***.***>:

> ***@***.**** approved this pull request.
> 
> LGTM 👍
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/home-assistant/core/pull/81149#pullrequestreview-1233481218], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AALDOPP6TT6KPBJD76MNOBDWQI55RANCNFSM6AAAAAARRIYCNE].
> You are receiving this because you authored the thread.[Tracking image][https://github.com/notifications/beacon/AALDOPOVGYJJKIFTMQMMA6LWQI55RA5CNFSM6AAAAAARRIYCNGWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSJQVXAE.gif]
>

ZephireNZ commented on homeassistant/components/electric_kiwi/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "domain": "electric_kiwi",
+  "name": "Electric Kiwi",
+  "config_flow": true,
+  "integration_type": "hub",
```
Just realised - should this be `service` instead?

From [the blog post](https://developers.home-assistant.io/blog/2022/10/24/hubs-devices-services-virtual-integrations/)

> A `hub` provides a gateway to multiple other devices or services. `service` and `device` are integrations that provide a single device or service per config entry.

mikey0000 replied:
You'll also need to add that to the flick electric integration :smile:

mikey0000 commented:
So in short yes it looks like it should be a service. Hopefully that can work with the energy stuff too. Will update.

3/01/2023 4:06:35 pm Brynley McDonald ***@***.***>:

> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In homeassistant/components/electric_kiwi/manifest.json[https://github.com/home-assistant/core/pull/81149#discussion_r1060255142]:
> 
>> @@ -0,0 +1,14 @@
> +{
> +  "domain": "electric_kiwi",
> +  "name": "Electric Kiwi",
> +  "config_flow": true,
> +  "integration_type": "hub",
> 
> Just realised - should this be *service* instead?
> 
> From the blog post[https://developers.home-assistant.io/blog/2022/10/24/hubs-devices-services-virtual-integrations/]
> 
> A *hub* provides a gateway to multiple other devices or services. *service* and *device* are integrations that provide a single device or service per config entry.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/home-assistant/core/pull/81149#pullrequestreview-1234062469], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AALDOPO2W23EVYQXWDHDHD3WQOJSXANCNFSM6AAAAAARRIYCNE].
> You are receiving this because you authored the thread.[Tracking image][https://github.com/notifications/beacon/AALDOPMIC4LCNPSF7KHYPDTWQOJSXA5CNFSM6AAAAAARRIYCNGWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSJRZGIK.gif]
>

mikey0000 commented:
I followed other integrations as examples (at the time), due to the fact it will provide multiple entities in the near future. So happy to take advice here on what is the correct integration type.

3/01/2023 4:06:35 pm Brynley McDonald ***@***.***>:

> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In homeassistant/components/electric_kiwi/manifest.json[https://github.com/home-assistant/core/pull/81149#discussion_r1060255142]:
> 
>> @@ -0,0 +1,14 @@
> +{
> +  "domain": "electric_kiwi",
> +  "name": "Electric Kiwi",
> +  "config_flow": true,
> +  "integration_type": "hub",
> 
> Just realised - should this be *service* instead?
> 
> From the blog post[https://developers.home-assistant.io/blog/2022/10/24/hubs-devices-services-virtual-integrations/]
> 
> A *hub* provides a gateway to multiple other devices or services. *service* and *device* are integrations that provide a single device or service per config entry.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/home-assistant/core/pull/81149#pullrequestreview-1234062469], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AALDOPO2W23EVYQXWDHDHD3WQOJSXANCNFSM6AAAAAARRIYCNE].
> You are receiving this because you authored the thread.[Tracking image][https://github.com/notifications/beacon/AALDOPMIC4LCNPSF7KHYPDTWQOJSXA5CNFSM6AAAAAARRIYCNGWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSJRZGIK.gif]
>

mikey0000 commented:
@frenck an email has been sent to hello@... from electric kiwi for supplying client_id and secret, would be good to understand what if any changes are required on my part to support this.

thank you.

mikey0000 commented:
@ZephireNZ need your approval for the workflow please. I've fixed up the tests and coverage etc should be good now.

mikey0000 commented:
I've refactored the sensor entity so more entities can be added and without additional API calls required, at least a good step forward before I PR the more complex entity for changing free power hour separately.

mikey0000 commented:
I've started to advance onto more entities in a seperate branch so this change doesn't become too huge.

mikey0000 commented:
Hopefully last commit, reduced the PR down to its smallest part required to be useful, which is why it has both select and two sensors. The select controls the two sensors for selecting the free hour of power, two sensors so users can create automations based on start and end time. FYI hopefully this still meets the checklist.

matthuisman commented:
Ideally you could have integration of a device reading your live kwh usage into HA and then selecting the best HOP at 11:45pm each night :)

mikey0000 commented:
> Ideally you could have integration of a device reading your live kwh usage into HA and then selecting the best HOP at 11:45pm each night :)

@matthuisman thats exactly what you can do now if you already have CT clamps on your power (how I'm using it now) I've moved further development into a HACS compatible repo https://github.com/mikey0000/ha-electric-kiwi where I'm tracking issues and feature requests, so feel free to add that as a feature. EK's power usage doesn't update in real time enough AFAIK to provide hourly KW/h usage. At least comparing it to my CT clamps it updates every day or so.

I'll open new PR's into core once this is merged.

If you wouldn't mind reviewing and approving if your happy, would help push this PR along. I also need to keep it small as per HA's new component rules/guides.

matthuisman commented on homeassistant/components/electric_kiwi/select.py:
think hop_select.py would be a better name?

ZephireNZ replied:
Nah the way the HA platform framework works, the file name has to be `select.py`

matthuisman replied:
understood, thanks :)

Lurchi70 commented on homeassistant/components/electric_kiwi/api/README.md:
```diff
@@ -0,0 +1,16 @@
+# electrickiwi.api
```
From my experience, it is easier to move the entire api code in a pypi package and referr to it in the manifest.yaml as requirement.  
This also removes non-home assistant specific code from the core codebase (beside the advantages, api specific code is not wanted in the code base) 
see https://developers.home-assistant.io/docs/api_lib_index

mikey0000 replied:
I will need to investigate that as its a common pattern I took out of other components including a google one. It also requires `from homeassistant.helpers import config_entry_oauth2_flow` The api.py is only for oauth a session.
I followed
https://developers.home-assistant.io/docs/api_lib_auth

mikey0000 replied:
I should add I did write an entire API client externally that is referenced in manifest.yaml

mikey0000 replied:
on a different note the issues with mypy is resolved now.

mikey0000 replied:
@Lurchi70 I don't think what I've done is wrong, I think you may be assuming I've got the API in the PR which I don't. Welcome feedback. thanks.

ZephireNZ replied:
The comment is on a super old commit (which does have the API code), that might be where the confusion is coming from.

Lurchi70 replied:
Correct. Sorry for the confusion. 
Hae seen that there was a lot of progres later.

mikey0000 commented:
I'll investigate that, it's a pattern common in home assistant already, hence why I did it. Pretty sure it's also specified in the auth token docs for developing integrations.

25/04/2023 6:54:31 pm Lurchi70 ***@***.***>:

> 
> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In homeassistant/components/electric_kiwi/api/README.md[https://github.com/home-assistant/core/pull/81149#discussion_r1176080904]:
> 
>> @@ -0,0 +1,16 @@
> +# electrickiwi.api
> 
> From my experience, it is easier to move the entire api code in a pypi package and referr to it in the manifest.yaml as requirement.
> This also removes non-home assistant specific code from the core codebase (beside the advantages, api specific code is not wanted in the code base)
> see https://developers.home-assistant.io/docs/api_lib_index
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/home-assistant/core/pull/81149#pullrequestreview-1399242366], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AALDOPJY65AGV7ABNGSDBELXC5YGLANCNFSM6AAAAAARRIYCNE].
> You are receiving this because you authored the thread.[Tracking image][https://github.com/notifications/beacon/AALDOPPHONY7VYN7SAYLGIDXC5YGLA5CNFSM6AAAAAARRIYCNGWGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSTM27H4.gif]
>

frenck commented on homeassistant/components/electric_kiwi/__init__.py:
```diff
@@ -0,0 +1,71 @@
+from .coordinator import ElectricKiwiHOPDataCoordinator
+
+PLATFORMS: list[Platform] = [
+    Platform.SENSOR,
+    Platform.SELECT,
```
Please reduce this PR to a single platform, as per:

<https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible>

And:

<https://developers.home-assistant.io/docs/review-process>

I would recommend keeping the sensor platform and remove the select platform for now.

frenck commented on homeassistant/components/electric_kiwi/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    )
+    hop_coordinator = ElectricKiwiHOPDataCoordinator(hass, ek_api)
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
```
```suggestion

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
```

frenck commented on homeassistant/components/electric_kiwi/__init__.py:
```diff
@@ -0,0 +1,71 @@
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        "ek_api": ek_api,
```
The coordinator already has the API available, why store it separately again?

mikey0000 replied:
my lack of time doing OOP and python is showing. Sorted

frenck commented on homeassistant/components/electric_kiwi/__init__.py:
```diff
@@ -0,0 +1,71 @@
+
+    # we need to set the client number and connection id
+    try:
+        await hass.data[DOMAIN][entry.entry_id]["ek_api"].set_active_session()
```
```suggestion
        await ek_api.set_active_session()
```

frenck commented on homeassistant/components/electric_kiwi/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    try:
+        await hass.data[DOMAIN][entry.entry_id]["ek_api"].set_active_session()
+        await hop_coordinator.async_config_entry_first_refresh()
+    except AuthException as err:
```
The coordinator call can't throw an `AuthException`? Why is it in this try/except clause?

mikey0000 replied:
set_active_session() could however very unlikely since we've already authed, removed

frenck commented on homeassistant/components/electric_kiwi/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+            for k, v in self.hop_intervals.intervals.items()
+        }
+
+    def get_selected_hop(self) -> Hop:
+        """Get currently selected hop."""
+        return self.data
```
This method seems unneeded, as it returns self.data, just use self.data in that case?

mikey0000 replied:
removed

frenck commented on homeassistant/components/electric_kiwi/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+HOP_SCAN_INTERVAL = timedelta(hours=2)
+
+
+class ElectricKiwiHOPDataCoordinator(DataUpdateCoordinator):
```
```suggestion
class ElectricKiwiHOPDataCoordinator(DataUpdateCoordinator[Hop]):
```

frenck commented on homeassistant/components/electric_kiwi/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "config_flow": true,
+  "dependencies": ["application_credentials"],
+  "documentation": "https://www.home-assistant.io/integrations/electric_kiwi",
+  "homekit": {},
```
```suggestion
```

frenck commented on homeassistant/components/electric_kiwi/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "integration_type": "hub",
+  "iot_class": "cloud_polling",
+  "requirements": ["electrickiwi-api==0.8.4"],
+  "ssdp": [],
+  "zeroconf": []
```
```suggestion
```

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+)
+
+
+class ElectricKiwiHOPEntity(CoordinatorEntity, SensorEntity):
```
```suggestion
class ElectricKiwiHOPEntity(CoordinatorEntity[ElectricKiwiHOPDataCoordinator], SensorEntity):
```

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+    ) -> None:
+        """Entity object for Electric Kiwi sensor."""
+        super().__init__(hop_coordinator)
+        self._hop_coordinator: ElectricKiwiHOPDataCoordinator = hop_coordinator
```
Why set this again? The `CoordinatorEntity` already has the `coordinator` property available?

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        self.connection_id = connection_id
+        self.entity_description = description
+        self._attributes = {
+            ATTR_ATTRIBUTION: ATTRIBUTION,
```
Attribution can be set via the attribution property instead.

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return "_".join(
```
We can just use an f-string instead.

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+            ]
+        )
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self.entity_description.name}"
```
This is already the default behavior?

```suggestion
```

mikey0000 replied:
didn't realise that, removing

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        return f"{self.entity_description.name}"
+
+    @property
+    def native_value(self) -> datetime | str | None:
```
How would this return an `str` or `None`?

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+            return value + timedelta(days=1)
+        return value
+
+    @property
+    def extra_state_attributes(self) -> dict[str, str]:
+        """Return the state attributes."""
+        return self._attributes
```
As per comment above, use the attribution method instead.

```suggestion
```

mikey0000 replied:
I've moved this to `_attr_attribution = ATTRIBUTION` I hope this is more acceptable

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        """Return the state attributes."""
+        return self._attributes
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Update attributes when the coordinator updates."""
+        self._attr_native_value = self.entity_description.value_func(
+            self._hop_coordinator.get_selected_hop()
+        )
+        super()._handle_coordinator_update()
```
It is already implementing the `native_property` method, why is the update handler implemented?

mikey0000 replied:
I'll commit a change that removes this and also tidies up the native_value function as well. I'm not sure on where to place the code that currently resides inside native_value so its a lone function, advice please. Will be pushing all these changes shortly

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,140 @@
+        )
+        super()._handle_coordinator_update()
+
+    async def async_update(self) -> None:
+        """Get the latest data from Electric Kiwi and updates the HOP."""
+        await super().async_update()
+        _LOGGER.debug("HOP data from sensor: %s", self._hop_coordinator.data)
```
What does this do?

frenck reviewed (changes requested):
Some initial comments to get the review process going.

../Frenck

mikey0000 commented:
Note, I've had to add # type: ignore[attr-defined] as mypy is incorrectly indicating that the model doesn't have the attributes but it does, welcome feedback or ways to resolve this better.

mikey0000 commented:
> Some initial comments to get the review process going.
> 
> ../Frenck

Have addressed all comments/code, not sure why github is still saying 1 change requested. No need to reply will wait.

mikey0000 commented on homeassistant/components/electric_kiwi/coordinator.py:
```diff
@@ -0,0 +1,74 @@
+
+    async def async_update_hop(self, hop_interval: int) -> Hop:
+        """Update selected hop and data."""
+        self.data = await self._ek_api.post_hop(hop_interval)
```
this looks like a better pattern

```

async def async_update_hop(self, hop_interval: int) -> Hop:
        """Update selected hop and data."""
        try:
            self.async_set_updated_data(await self._ek_api.post_hop(hop_interval))
        except AuthException as auth_err:
            raise ConfigEntryAuthFailed from auth_err
        except ApiException as api_err:
            raise UpdateFailed(
                f"Error communicating with EK API: {api_err}"
            ) from api_err
```

mikey0000 committed:
remove unneeded scopes from test consts

mikey0000 committed:
move fixtures to conftest

mikey0000 committed:
move consts to conftest

mikey0000 committed:
Merge remote-tracking branch 'nabucasa/dev' into feature/electric-kiwi

mikey0000 committed:
Merge branch 'dev' into feature/electric-kiwi

frenck commented on homeassistant/components/electric_kiwi/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    )
+    hop_coordinator = ElectricKiwiHOPDataCoordinator(hass, ek_api)
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        "hop_coordinator": hop_coordinator,
+    }
```
We don't have to create a dictionary if there is a single entry in it.

```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = hop_coordinator
```

Additionally, this must be moved down until after the first refresh has completed. Otherwise we may end up with object in hass data that aren't viable (e.g., if `ConfigEntryNotReady was thrown).

mikey0000 replied:
FYI next PR will contain an additional coordinator. My HACS version already does. It is a bit of work to change this only to change it back on the next PR. I'll move it, but can I keep the dictionary please?

frenck replied:
Let's not add things now that aren't needed. If a follow up requires more to be stored, we can do that at that point. In this PR it has no use. Please remove the mapping.

frenck commented on homeassistant/components/electric_kiwi/coordinator.py:
```diff
@@ -0,0 +1,81 @@
+        filters the intervals to remove ones that are not active
+        """
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
```
Note something we should document here, that is a general given.

```suggestion
```

frenck commented on homeassistant/components/electric_kiwi/coordinator.py:
```diff
@@ -0,0 +1,81 @@
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(60):
+                if len(self.hop_intervals.intervals) == 0:
```
```suggestion
                if not len(self.hop_intervals.intervals):
```

mikey0000 replied:
can't do according to the linter `not len()` so refactoring this type since its a dictionary.

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,122 @@
+HOP_SENSOR_TYPE: tuple[ElectricKiwiHOPSensorEntityDescription, ...] = (
+    ElectricKiwiHOPSensorEntityDescription(
+        key=ATTR_EK_HOP_START,
+        name="Hour of free power start",
```
Add a `translation_key` and move the actual name into `strings.json`

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    ),
+    ElectricKiwiHOPSensorEntityDescription(
+        key=ATTR_EK_HOP_END,
+        name="Hour of free power end",
```
Add a `translation_key` and move the actual name into `strings.json`

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,122 @@
+        """Entity object for Electric Kiwi sensor."""
+        super().__init__(hop_coordinator)
+
+        self.customer_number = self.coordinator._ek_api.customer_number
+        self.connection_id = self.coordinator._ek_api.connection_id
```
```suggestion
        self._attr_unique_id = f"{coordinator._ek_api.customer_number}_{coordinator._ek_api.connection_id}_{description.key}"
```

frenck commented on homeassistant/components/electric_kiwi/sensor.py:
```diff
@@ -0,0 +1,122 @@
+        self.connection_id = self.coordinator._ek_api.connection_id
+        self.entity_description = description
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return f"{self.customer_number}_{str(self.connection_id)}_{self.entity_description.key}"
+
```
```suggestion
```

mikey0000 committed:
Update homeassistant/components/electric_kiwi/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

mikey0000 committed:
Update homeassistant/components/electric_kiwi/sensor.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

mikey0000 committed:
Update homeassistant/components/electric_kiwi/coordinator.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

mikey0000 committed:
Update homeassistant/components/electric_kiwi/coordinator.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

mikey0000 committed:
remove dictionary and tidy up strings as well as hop intervals logic

mikey0000 committed:
fix sensors to correct start and end respectively

frenck reviewed (approved):
Looks good to me. Thanks @mikey0000 👍 

../Frenck

mikey0000 commented:
> Looks good to me. Thanks @mikey0000 +1
> 
> ../Frenck

Thank you! Thank you! Thank you! So happy! :smile:

mikey0000 commented:
@frenck sorry to have to @ you but https://github.com/home-assistant/core/pull/97202
the translation keys require the attribute set, so here is the bugfix for it, the translations weren't working.
